Static tramp v5

madvenka786 opened the pull request:
Fixed a couple of build problems. Also excluded cygwin from static trampolines. This is ready for review as soon as the CI tests pass.

atgreen commented:
Can you comment on the thread-safety of your patch?  ffi_tramp_init in particular.   We punt responsibility for thread-safety to the calling program, but we should at least add a note to the thread-safety section of the documentation.

atgreen commented:
If I'm reading this patch right, it will try to use the static trampolines and fall back to the dynamic method if that fails.   Why would static trampolines every not be available on a supported platform.  I'm wondering if we can simply not include the old dynamic method.

madvenka786 commented:
ffi_tramp_init() is a (static) internal function that is only called by ffi_tramp_is_supported() and ffi_tramp_alloc(). In both cases, the global static trampoline lock is acquired via ffi_tramp_lock(). In fact, all static trampoline operations acquire the lock. So, they are thread-safe.

madvenka786 commented:
On the fallback question - the issue is in ffi_tramp_init_os (). This is the OS-specific initialization function. It needs to call OS-specific functions which can potentially fail. libffi must be resilient to that. E.g., let us consider loading the trampoline code in a temporary file and mapping it. Today, this is allowed. Tomorrow, an OS may require a valid file signature to allow mmap(file) with EXEC permissions. That will not work on temporary files.

madvenka786 commented:
I have addressed all of the comments so far. Is there anything else I should address? If not, can these changes be merged into libffi?
Please let me know. Thanks!

atgreen commented:
Thank you.  I have merged this great improvement to libffi.   The testing all looks good, so I hope there will be no problems.

madvenka786 commented:
Thank you so much!

Madhavan

On 3/5/21 10:09 AM, Anthony Green wrote:
> Thank you. I have merged this great improvement to libffi. The testing all looks good, so I hope there will be no problems.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/libffi/libffi/pull/624#issuecomment-791517222>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AR54PMXQDIYGGJP6Z3IRRMTTCD62DANCNFSM4XGXWZWQ>.
>

kohtala commented on src/tramp.c:
```diff
@@ -0,0 +1,729 @@
+  uintptr_t addr = (uintptr_t) tramp_globals.text;
+  int nfields, found;
+
+  snprintf (file, PATH_MAX, "/proc/%d/maps", getpid());
```
Hi. I'm curious if there is some case where /proc/self/maps does not work here. If there is, a comment would have been useful.

madvenka786 replied:
maps has been around for a while and has remained stable AFAICT. However, I have assumed that it can fail. If it does, the code falls back to using a temporary file to map the trampoline table.

kohtala replied:
I was curious enough to actually trace /proc/pid/maps to first appear in Linux 0.99.13k, but at that time it did not have the file name column. The file name column was at least in 2.4.12 if not even earlier. But I was thinking about use of `snprintf` and `getpid` being useless because can't you just open /proc/**self**/maps. Unless I miss something. /proc/self symlinks to getpid() so it should work the same. As it works the same, this is nothing major, just avoids any chance of mistakes with snprintf to stack. I checked /proc/self too, and it appeared in Linux-0.99 back in 1992.

madvenka786 replied:
I agree. I could/should have used /proc/self/maps. That would have been more elegant.

kohtala replied:
Ok. Thanks. My major concern was if there is something I did not know and you could tell me why it indeed makes sense to use getpid and snprintf.

kohtala commented:
> E.g., let us consider loading the trampoline code in a temporary file and mapping it. Today, this is allowed. Tomorrow, an OS may require a valid file signature to allow mmap(file) with EXEC permissions. That will not work on temporary files.

To add, it is not just about signatures. Today, some security conscious have all their executable files on read only filesystems. If any mount is writable, it can be noexec mounted and there is no executable mmap from it. Some mount only /tmp noexec.

sempervictus commented:
Neat, thank you for the effort.
Far as `/proc` restrictions go - the Grsecurity patches have offered those for a long time, though up until pretty recently they could cause some serious confusion in userspace. LSMs and namespaces (systemd comes to mind, ditto container engines) may apply similar restrictions, so might be worthwhile to create a super-restricted systemd service using the new code to make sure it doesn't trip over itself when using foreign interfaces. If systemd's involved, humans will be likely just blame those hapless people since its basically fait accompli that they're at fault for everything around init nowadays.
For the LSM bit, i'd be most concerned with the Android SELinux implementation - they're starting to get up there with GrapheneOS as time goes on. Probaifbly merits testing as well - orgs like No Such Animal are heavy users, would not be great to break their stuff in a major foundational library.
On the subject to writing out mmap-able files, this is (IMO) a great idea since today FFI trips on RWX defenses (or did before this), and as more systems implement W^X, this may become more and more of a viable use-case. The `noexec` mount thing is becoming more common too, but that's actually a good thing as it should eventually lead to some new mountpoint for this specific use-case (some `tmpfs` style mounts per-user/per-pid to have semi-scary code run in a way other UIDs/PIDs cant reach). The discussion on signed ELF is quite old, and may bear fruit at some point, but probably not until were in next-gen OS territory at this rate.

madvenka786 commented:
Hi Anthony,

How are you?

Can you tell me what version of libffi this landed in? Our Yocto needs to be updated to the right version of libffi.

Thanks in advance for the information.

Madhavan

On 3/5/21 10:07, Anthony Green wrote:
> Merged #624 <https://github.com/libffi/libffi/pull/624> into master.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/libffi/libffi/pull/624#event-4414317729>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AR54PMXWZNXABYVQDV5QM73TCD6VHANCNFSM4XGXWZWQ>.
>

madvenka786 commented:
Thanks!

Madhavan

On 3/7/22 08:00, Anthony Green wrote:
> Hi Madhavan,
> 
> This was included in version 3.4.2.
> 
> AG
> 
> 
> On Wed, Mar 2, 2022 at 2:04 PM Madhavan T. Venkataraman <
> ***@***.***> wrote:
> 
>> Hi Anthony,
>>
>> How are you?
>>
>> Can you tell me what version of libffi this landed in? Our Yocto needs to
>> be updated to the right version of libffi.
>>
>> Thanks in advance for the information.
>>
>> Madhavan
>>
>> On 3/5/21 10:07, Anthony Green wrote:
>> > Merged #624 <https://github.com/libffi/libffi/pull/624> into master.
>> >
>> > —
>> > You are receiving this because you authored the thread.
>> > Reply to this email directly, view it on GitHub <
>> https://github.com/libffi/libffi/pull/624#event-4414317729>, or
>> unsubscribe <
>> https://github.com/notifications/unsubscribe-auth/AR54PMXWZNXABYVQDV5QM73TCD6VHANCNFSM4XGXWZWQ
>> >.
>> >
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/libffi/libffi/pull/624#issuecomment-1057278229>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAAV7CK7SHDA5APZRV47K7TU563URANCNFSM4XGXWZWQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>>
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/libffi/libffi/pull/624#issuecomment-1060717158>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AR54PMRMUVTYPU6XPDP6DPLU6YDXDANCNFSM4XGXWZWQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
> You are receiving this because you authored the thread.Message ID: ***@***.***>
>
